### PR TITLE
fix: isolate Builder snapshots

### LINKED ISSUE #494: Expose readonly in-progress Builder values
**Milestone:** Phase 13 Extraction Workflows — M1 (D1.2)
**PRD:** `docs/project/PRD/phase-13-extraction-workflows/PRD.md` §§4 and 6
**Architecture:** `docs/project/PRD/phase-13-extraction-workflows/ARCH.md` §3.4
**Dogfooding:** Not applicable for this product-code deliverable
**Delivery plan:** `docs/project/PRD/phase-13-extraction-workflows/delivery.md` D1.2

## Problem

> Riviere extraction today is a sequence of manual CLI commands.

Later stages and public transition results need exact accumulated graph values before final build, but Builder has no supported readonly component, Link, or external-Link views. This slice matters because D1.2 is an explicit dependency in the approved delivery sequence and an incomplete boundary would push design decisions into later implementation tickets.

## Solution

Expose immutable snapshots of in-progress Builder values without exposing mutable collections or constructing a completed query model. This ticket delivers only this approved slice and leaves neighbouring capabilities to their named deliverables.

### Full architecture context

Read `docs/project/PRD/phase-13-extraction-workflows/ARCH.md` for the full approved architecture context and scope. This issue is not a replacement for `ARCH.md`.

### Target components, boundaries, and responsibilities

- Builder remains the source of graph identity and merge truth.
- Views support Project behaviour and transition snapshots.
- Existing occurrence and warning semantics remain unchanged.

### Code/config/flow from ARCH.md

```typescript
builder.components(): readonly Component[]
builder.links(): readonly Link[]
builder.externalLinks(): readonly ExternalLink[]
```

### Role and location decisions

| Proposed Element | Kind | Role | Sublocation | Confidence | Notes |
| --- | --- | --- | --- | --- | --- |
| RiviereBuilder | class | value-object | riviere-builder published-language | High | Mutable graph-building value with readonly snapshots |

### Structural concerns and forbidden alternatives

- Do not return mutable internal arrays/maps.
- Do not use `builder.build()` for invalid intermediate-state inspection.
- Do not extend `RiviereQuery`.

### Design challenge instruction

> If implementation reveals the agreed design is impractical, incomplete, or sub-optimal, stop and push back. Do not silently implement a different design.

### Product

- Callers receive exact accumulated components, internal Link occurrences, and external Links before build.
- Returned values cannot mutate Builder state.

### Design

- No `RiviereQuery` dependency or query-package cycle exists.
- Builder remains a mutable `value-object`; no identity-bearing role is invented.
- Internal Link source occurrences and external-Link dedup warnings remain intact.

### Quality

- Focused tests prove immutability and exact values; existing 80,000-item performance coverage remains passing.

### Change areas

- `packages/riviere-builder/published-language/`
- Builder unit, regression, and performance tests

### Relevant repository rules

- `docs/conventions/software-design.md` SD-001 (fail fast), SD-002/SD-003 (exhaustive type-safe states), SD-004 (dependency injection), SD-005 (intention-revealing names), and SD-021 (YAGNI) apply where this ticket touches code.
- `docs/conventions/testing.md` TS-001 through TS-004 require outcome-named tests with exact value assertions and one concept per test; TS-008/TS-014 require relevant boundary cases.

### Relevant architecture memories

- `project-memory/architecture/memories/repository-loading-must-load-full-aggregate.md`: repository criteria describe full aggregate state, not the operation that follows loading.
- `project-memory/architecture/memories/ddd-modelling-trade-offs-from-riviere-workflows.md`: Project owns empty-start rebuild meaning and Workflow owns the stage lifecycle; repositories and use cases do not become the engine.
- `project-memory/architecture/memories/rejected-workflow-use-case-dumping-case-study.md`: do not dump stage progression or low-level collaborators into a command use case.

### Lint, role-check, and test expectations

Use role and lint failures to improve the design; do not add ignores, exclusions, assertions, optional production dependencies, or coverage workarounds. Preserve exhaustive unions and avoid the TypeScript `in` operator.

## Dependencies

- #491 (D1.0)

## How to verify

- Run Builder tests; exact readonly values, mutation attempts, occurrences, dedup warnings, and upserts pass.
- Run existing Builder performance tests; no regression is introduced.
- Run `pnpm verify`.

## Things this ticket must not do

- Do not return mutable internal arrays/maps.
- Do not use `builder.build()` for invalid intermediate-state inspection.
- Do not extend `RiviereQuery`.

### PR DESCRIPTION
**Pull request by NTCoding**
## Description

Closes #494.

## Linked Issue

Closes #494

## What Problem Does This PR Solve?

Builder snapshots could expose mutable metadata references.

## Acceptance Criteria

Snapshots are detached and unsupported metadata fails fast.

## Key Changes

Adds detached graph snapshots and metadata validation.

## Notable Architectural Changes / Impact

Snapshot construction is a Builder value object.

## Validation

pnpm verify; required reviews passed.

## Notes

No additional notes.

## Summary by CodeRabbit

- **New Features**
  - Pull requests now target the recorded feature branch.
  - Pull-request descriptions must contain at least 100 characters, with clear validation messages.
  - GitHub feedback is now fully paginated for more complete reviews.
  - Feedback fixes can be pushed and rechecked through the workflow.
  - Graph data can be inspected through isolated snapshots, preserving complex metadata safely.

- **Bug Fixes**
  - Missing feature branches now produce a clear creation error.
  - CodeRabbit rate limits now block the workflow with guidance to retry later.
  - Verified feedback transitions the workflow into reflection.

### CODE REVIEW

File: packages/dev-workflow-v2/use-cases/src/infra/external-clients/git/git-client.ts [hunk 1]
```diff
@@ -40,6 +40,15 @@ export function readGitRepositoryStatus(
   }
 }
 
+/** @riviere-role external-client-service */
+export function pushGitBranch(
+  branch: string,
+  gitBinary = 'git',
+  executeGit: GitExecutor = defaultGitExecutor,
+): void {
+  executeGit(gitBinary, ['push', '--set-upstream', 'origin', branch])
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/dev-workflow-v2/use-cases/src/infra/external-clients/git/git-client.ts]: ✅ **Fixed**: PR creation now passes `--head` and no longer forces a push to `origin`.

File: packages/dev-workflow-v2/use-cases/src/infra/external-clients/github/create-pull-request.ts [hunk 2]
```diff
@@ -33,8 +35,10 @@ class PullRequestCreationOutputError extends Error {
 /** @riviere-role external-client-service */
 export function createGithubPullRequestClient(
   runGh: GhRunner,
+  pushBranch: PushBranch,
 ): (request: GithubPullRequestCreationInput) => GithubPullRequest {
   return (request: GithubPullRequestCreationInput): GithubPullRequest => {
+    pushBranch(request.branch)
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/dev-workflow-v2/use-cases/src/infra/external-clients/github/create-pull-request.ts]: ✅ **Fixed**: PR creation now passes the recorded branch through `--head`.

File: packages/riviere-builder/published-language/src/published-language/riviere-builder.ts [hunk 3]
```diff
@@ -425,7 +434,7 @@ export class RiviereBuilder {
 
   /** @returns The current graph encoded as JSON. */
   serialize(): string {
-    return JSON.stringify(this.serializedGraph(), null, 2)
+    return JSON.stringify(this.inspectionGraph(), null, 2)
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/riviere-builder/published-language/src/published-language/riviere-builder.ts]: ✅ **Fixed**: `serialize()` now serializes the validated detached published graph.

**Comment by NTCoding** [comment]
@coderabbitai review

File: packages/riviere-builder/published-language/src/published-language/riviere-graph-snapshot.ts [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+  clones.set(publishedDate, snapshotDate)
+  copyPublishedProperties(publishedDate, snapshotDate, clones)
+  return snapshotDate
+}
+
+function cloneRegularExpression(
+  publishedExpression: RegExp,
+  clones: WeakMap<object, object>,
+): RegExp {
+  const snapshotExpression = new RegExp(publishedExpression.source, publishedExpression.flags)
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/riviere-builder/published-language/src/published-language/riviere-graph-snapshot.ts]: ✅ **Fixed**: RegExp snapshot cloning now uses `structuredClone`, avoiding dynamic pattern construction while retaining `lastIndex` and own-property coverage.

**Comment by NTCoding** [comment]
@coderabbitai review

File: apps/eclair/src/features/full-graph/entrypoint/FullGraphPage.spec.tsx [hunk 5]
```diff
@@ -524,26 +524,16 @@ describe('FullGraphPage', () => {
...
 
       await act(async () => {
-        vi.advanceTimersByTime(199)
-      })
-
-      expect(screen.getByTestId('graph-tooltip')).toBeInTheDocument()
-
-      await act(async () => {
-        vi.advanceTimersByTime(1)
+        await new Promise((resolve) => setTimeout(resolve, 250))
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/eclair/src/features/full-graph/entrypoint/FullGraphPage.spec.tsx]: ✅ **Fixed**: Restored fake timers and assertions that the tooltip remains visible at 199ms and is removed at 200ms.

File: packages/dev-workflow-v2/domain-model/src/domain/coderabbit-feedback-verification.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      type: 'retry',
+      consecutiveCleanPolls: 0,
+    }
+  }
+
+  const nextConsecutiveCleanPolls = verification.clean ? consecutiveCleanPolls + 1 : 0
+  if (
+    verification.clean &&
+    nextConsecutiveCleanPolls < REQUIRED_CONSECUTIVE_CLEAN_CODERABBIT_POLLS &&
+    attemptsRemaining > 1
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/dev-workflow-v2/domain-model/src/domain/coderabbit-feedback-verification.ts]: ✅ **Fixed**: Verification now requires two consecutive clean polls; a final incomplete confirmation produces timeout rather than verified.

File: tools/dev-workflow-v2/states/awaiting_pr_feedback.md [hunk 7]
```diff
@@ -13,11 +13,11 @@ When feedback is available, the workflow automatically transitions to the next s
...
 
-If CodeRabbit feedback appears and is not clean, the workflow transitions directly to `ADDRESSING_FEEDBACK`, where feedback is re-checked against live GitHub state before REVIEWING can resume. If the wait times out or feedback fetch fails, the workflow transitions to `BLOCKED`.
+If CodeRabbit reports that its review is rate limited, the workflow immediately transitions to `BLOCKED` and tells the user to wait for the rate limit to reset before returning to `AWAITING_PR_FEEDBACK`. Otherwise, if CodeRabbit feedback appears and is not clean, the workflow transitions directly to `ADDRESSING_FEEDBACK`, where feedback is re-checked against live GitHub state before REVIEWING can resume. If the wait times out or feedback fetch fails, the workflow transitions to `BLOCKED`.
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply tools/dev-workflow-v2/states/awaiting_pr_feedback.md]: ✅ **Fixed**: Awaiting-feedback instructions now name REFLECTING as the post-verification state.

File: packages/riviere-builder/published-language/src/published-language/builder-snapshots.spec.ts [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+    })
+  })
+
+  it('preserves regular expression metadata when a returned value is changed', () => {
+    const builder = createSnapshotBuilder()
+    const matcher = /order/gi
+    matcher.lastIndex = 4
+    addSnapshotPolicy(builder, { matcher })
+    const returnedMatcher = regularExpressionMetadata(policyMetadata(builder, 'matcher'))
+    returnedMatcher.lastIndex = 1
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/riviere-builder/published-language/src/published-language/builder-snapshots.spec.ts]: ✅ **Fixed**: The snapshot test now asserts cloned RegExp lastIndex is 4 before changing it.

File: packages/dev-workflow-v2/use-cases/src/features/workflow/adapters/github/workflow-pull-request-feedback-reader.spec.ts [hunk 9]
```diff
@@ -4,6 +4,7 @@ import { createWorkflowPullRequestFeedbackReader } from './workflow-pull-request
 it('translates GitHub feedback into workflow feedback', () => {
   const readFeedback = createWorkflowPullRequestFeedbackReader(() => ({
     coderabbitReviewSeen: true,
+    coderabbitRateLimited: false,
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/dev-workflow-v2/use-cases/src/features/workflow/adapters/github/workflow-pull-request-feedback-reader.spec.ts]: ✅ **Fixed**: Added adapter coverage that proves coderabbitRateLimited true is preserved.

File: packages/dev-workflow-v2/use-cases/src/infra/external-clients/github/get-pr-feedback.ts [hunk 10]
```diff
@@ -71,22 +87,35 @@ export function createGithubPullRequestFeedbackClient(
...
+          (review) =>
+            isCodeRabbitAuthor(review.author?.login) && indicatesCodeRabbitRateLimit(review.body),
+        ) ||
+        threads.some((thread) =>
+          thread.comments.some(
+            (comment) =>
+              isCodeRabbitAuthor(comment.author?.login) &&
+              indicatesCodeRabbitRateLimit(comment.body),
+          ),
+        ),
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/dev-workflow-v2/use-cases/src/infra/external-clients/github/get-pr-feedback.ts]: ✅ **Fixed**: Added paginated retrieval for reviews, threads, and comments. Rate-limit detection now uses the newest active CodeRabbit feedback and excludes resolved or outdated history.

File: tools/dev-workflow-v2/src/shell/cli.ts [hunk 11]
```diff
@@ -32,7 +32,7 @@ const routes = createWorkflowRoutes({
   parseStringArguments,
 })
 const bashForbidden = {
-  commands: ['git push', 'gh pr'],
+  commands: ['gh pr'],
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply tools/dev-workflow-v2/src/shell/cli.ts]: ✅ **Fixed**: Direct branch pushes remain globally forbidden. The new workflow operation sends only origin and the recorded feature branch during ADDRESSING_FEEDBACK.

File: apps/eclair/src/features/full-graph/entrypoint/FullGraphPage.filters.spec.tsx [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+      await user.click(domainCheckbox)
+      expect(screen.queryByTestId('focused-domain-banner')).toBeNull()
+    })
+
+    it('passes focusedDomain prop to ForceGraph when domain selected', async () => {
+      const user = userEvent.setup()
+      renderFullGraphPage()
+      await user.click(screen.getByTestId('filter-toggle'))
+      await user.click(screen.getByTestId('domain-checkbox-orders'))
+      expect(screen.getByTestId('force-graph-container')).not.toBeNull()
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/eclair/src/features/full-graph/entrypoint/FullGraphPage.filters.spec.tsx]: ✅ **Fixed**: The ForceGraph fixture now exposes focusedDomain through its rendered mock, and the filter test asserts the selected value is orders.

File: packages/dev-workflow-v2/use-cases/src/infra/external-clients/github/get-pr-feedback-pagination.spec.ts [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+            nodes: [],
+            pageInfo: NO_NEXT_PAGE,
+          },
+        },
+      },
+    },
+  })
+  const runGh = vi.fn().mockReturnValue(REPO_INFO)
+  runGh.mockReturnValueOnce(REPO_INFO).mockReturnValueOnce(rateLimited)
+  runGh.mockReturnValueOnce(REPO_INFO).mockReturnValueOnce(successful)
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply packages/dev-workflow-v2/use-cases/src/infra/external-clients/github/get-pr-feedback-pagination.spec.ts]: ✅ **Fixed**: The successful fixture now contains an earlier rate-limit review and a newer successful review, and a separate newest-first rate-limit case preserves full branch coverage.

**Comment by NTCoding** [comment]
@coderabbitai review
